wremnants/CardTool.py

```diff
@@ -420,15 +423,94 @@ def checkSysts(self, var_map, proc, thresh=0.25):
+                    for iy in range(len(dcbb[k]["edges"][1])-1):
+                        systNamesTmp.extend([f"{x}_{decorrSystLabel[0]}{ix}{decorrSystLabel[1]}{iy}" for x in systNames])
+            else:
+                for ik in range(len(dcbb[k]["edges"])-1):
+                    systNamesTmp.extend([f"{x}_{decorrSystLabel}{ik}" for x in systNames])
+            return systNamesTmp
+            #logger.error(f"systNames = {systNames}")
+
+    def makeDecorrelatedSystHistograms(self, h, hnomi, name, decorrByBinDict):
+        # s = hist.tag.Slicer()
+        # TODO: using an alternative version using boost directly may be faster, but for now this is just for testing
+        # decorrByBinDict is a dictionary as
+        # decorrByBinDict = {"x": {"label" : "eta",
+        #                          "edges" : [round(-2.4+i*0.4,1) for i in range(13)],}
+        #                   }
+        # for a 2D decorrelation the values of each key are arrays of size 2
+        # decorrByBinDict = {"xy": {"label" : ["eta", "pt"],
+        #                          "edges" : [ [etaEdges], [ptEdges] ]}
+        #                   }
+        # TODO: could also do charge decorrelation by using the z axis if present
+        ret = {}
+        hnomiroot = narf.hist_to_root(hnomi)
+        hsystroot = narf.hist_to_root(h)
+        for ax in decorrByBinDict.keys():
+            decorrDict = decorrByBinDict[ax]
+            decorrSystLabel = decorrDict["label"]
+            if ax != "xy":
+                for ibin in range(len(decorrDict["edges"]) -1):
+                    upDown = "Up" if name.endswith("Up") else "Down" if name.endswith("Down") else ""
+                    newname = name.rstrip(upDown)
+                    newname = f"{newname}_{decorrSystLabel}{ibin}{upDown}"
+                    logger.warning(f"Decorrelating syst {name} by {ax} bins: preparing new histogram {newname}")
+                    # FIXME: do not assume we can only have eta and pt axes (or eta-pt when implemented)
+                    # bin ID are retrieved adding some small constants to bin edges, low/high edge is increased/decreased
+                    if ax == "x":
+                        xbinLow = hnomiroot.GetXaxis().FindFixBin(decorrDict["edges"][ibin]+0.001)
+                        xbinHigh = hnomiroot.GetXaxis().FindFixBin(decorrDict["edges"][ibin+1]-0.001)
+                        ybinLow = 1
+                        ybinHigh = hnomiroot.GetNbinsY()
+                    elif ax == "y":
+                        xbinLow = 1
+                        xbinHigh = hnomiroot.GetNbinsX()
+                        ybinLow = hnomiroot.GetYaxis().FindFixBin(decorrDict["edges"][ibin]+0.001)
+                        ybinHigh = hnomiroot.GetYaxis().FindFixBin(decorrDict["edges"][ibin+1]-0.001)
+                    hsystrootDecorr = copy.deepcopy(hnomiroot.Clone(newname))
+                    ROOT.wrem.fillTH3fromTH3part(hsystrootDecorr, hsystroot,
```

Comment:
Do we need a C++ implementation of this or can this also be done with slicing the boost histogram? Maybe boost histogram can do it more efficient.

Reply:
I just wanted to use root because I control it better than boost, and this function was already in use elsewhere so it was promptly available. For the cases where it is currently used speed is not an issue.
In the future we can foresee a boost implementation